pkg/database/database.go

```diff
@@ -107,39 +127,81 @@ func (db *Database) QueryNamespacedResourceByName(ctx context.Context, kind, api
 	}
 }
 
-func (db *Database) performResourceQuery(ctx context.Context, query string, args ...string) ([]*unstructured.Unstructured, string, string, error) {
-	castedArgs := make([]interface{}, len(args))
-	for i, v := range args {
-		castedArgs[i] = v
-	}
+func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
 
-	rows, err := db.db.QueryContext(ctx, query, castedArgs...)
+	strQueryPerformer := newQueryPerformer[string](db.db)
+	if kind == "Pod" {
+		pq := db.newParamQuery(db.info.GetLogURLsByPodNameSQL())
+		pq.addStringParams(apiVersion, namespace, name)
+		return strQueryPerformer.performQuery(ctx, pq)
+	}
+	var uuids []string
+	var uuid string
+	err := db.db.QueryRowContext(ctx, db.info.GetUUIDSQL(), kind, apiVersion, namespace, name).Scan(&uuid)
 	if err != nil {
-		return nil, "", "", err
+		return nil, err
+	}
+	uuids = append(uuids, uuid)
+	var podUuids []string
+	type uuidKind struct {
+		Uuid string `json:"uuid"`
+		Kind string `json:"kind"`
+	}
+	// This loops for each jump existing in the ownerReference chain from the first resource until the pod
+	// The slice uuids is modified inside the loop
```

Comment:
Loop counters should generally not be modified within the loop. This tells me something is wrong.
```suggestion
	pq := db.newParamQuery(db.info.GetLogURLsSQL())
	pq.addStringArrayParam(podUuids)
	return strQueryPerformer.performQuery(ctx, pq)
```

Reply:
ok, so look at it in a different way, note that golang doesn't have the `while` reserve word, `while` is performed with the `for`. So, I am doing things `while` I still have `uuid`s of  resources different of a `Pod`, but I am not iterating over an array/slice. In other words, I am not modifying the counter of a loop, I am doing things while I still have uuids, and I am removing uuids in the process and adding new ones

Reply:
You seem intent on ignoring my suggestion to use recursion to solve this problem, and I'm not sure why.

The fact that what you are trying to do isn't supported easily in golang indicates to me that you should consider other options.  What I am proposing works and I think is far more readable. I have been looking at this code for the past couple of days and I find it very confusing.  The fact that you have to explain in such detail also indicates that it's either overly complex or just unclear.

You seem to be focused on making the fewest number of calls possible, which is an admirable goal. But to me the tradeoff with the complexity and the lack of readability of the code makes it not worth it. Here are two examples I can cite:

1. The first call to special-case if the request is for a `Pod`.  Yeah, you can do that and you make one less call. But to me a more general solution that just always gets the uuid and then goes from there is easier to read and more concise. Everything is handled the same.  I'm not sure the tradeoff is worth it for the number of times logs will be requested in general versus the number of times logs are requested specifically for a `Pod`.
2. Batching uuids together for calls. I understand this a little more, but again I'm not sure it's worth the tradeoff. It makes no difference in your implementation, and it's clear why you are doing it. But I can tell you I didn't notice this batching the first N times I read the code, and as  a result I thought the code was broken, which again points to readability. Again, I question whether this is worth it, but recursion can also support doing it that way as well.

So I will meet you halfway and give you #2 but not number #1.  Here's yet another solution with that in mind:

```suggestion
type uuidKind struct {
	Uuid string `json:"uuid"`
	Kind string `json:"kind"`
}

func (db *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace, name string) ([]string, error) {
        []string PodUuids
        string owner
	err := db.db.QueryRowContext(ctx, db.info.GetUUIDSQL(), kind, apiVersion, namespace, name).Scan(&owner)
	if err != nil {
		return nil, err
	}
        podUuids, err := getPodUUIDs(ctx, [ owner ], podUuids)
	pq := db.newParamQuery(db.info.GetLogURLsSQL())
	pq.addStringArrayParam(podUuids)
	return strQueryPerformer.performQuery(ctx, pq)
}

func (db *Database) getPodUUIDs(ctx context.Context, []string owners, []string podUuids) ([]string, error) {
	if len(uuids) == 0 {
	    return podUuids, nil
        } else {
		pq := db.newParamQuery(db.info.GetOwnedResourcesSQL())
		pq.addStringArrayParam(owners)
		parsedRows, err := newQueryPerformer[uuidKind](db.db).performQuery(ctx, pq)
		if err != nil {
			return nil, err
		}

                var uuids string[]
                for _, row := range parsedRows {
			if row.Kind == 'Pod' {
				podUuids = append(podUuids, row.Uuid)
			} else {
				uuids = append(uuids, row.Uuid))
			}
                }
                return getPodUuids(ctx, uuids, podUuids)
        }
}
```
I don't like it as well as my previous solution, and I think it's uglier because it does the batching as you were. But I still think it's much more concise and readable that what you have.

> In other words, I am not modifying the counter of a loop, I am doing things while I still have uuids, and I am removing uuids in the process and adding new ones

I see your point here, but I think it's semantics. You aren't actually touching `i`. You are however assigning `i` a value rather than the normal increment.  It looks odd.

As I said, I don't like it, but this is even better IMO. You don't need `i` at all:
```
	for ; len(uuids) > 0;) {
```

Reply:
> You seem intent on ignoring my suggestion to use recursion to solve this problem, and I'm not sure why.

I am sorry, it wasn't my intent to ignore you and I had a comment written these days that I thought I sent but I didn't press the "Comment" button in Github -.-'
I'll try to address the points that I did there and also what you mention in here

> The fact that what you are trying to do isn't supported easily in golang

I am not sure why you are saying this, it is supported, it's just the way golang handles the `while` of other languages.

> You seem to be focused on making the fewest number of calls possible

Indeed I am. 
Let's say you have a large complex PipelineRun, with lots of TaskRuns, let's say 15. 
If I batch the uuids, I will only do 2 queries to get the owned resources, if I go one by one as you suggested at some point, you will have 16 queries, one per PipelineRun and one per TaskRun to find its pods. 
I really believe we should prioritize optimization in the number of queries for this use case.

> So I will meet you halfway and give you `#2` but not `#1`

I could live with this, even though it makes more sense to me trying to optimize the number of queries in general.

> Regarding recursion

What I mentioned in my lost comment was that I could do recursion while keeping the logic that I have. My idea was actually pretty similar to your current suggestion but I was going to say let's do it in a separate MR.
One other thing that I was considering for the recursion is to make it a [tail-recursion](https://medium.com/@meeusdylan/tail-recursion-in-go-fb5cf69a0f26) as it is more efficient. 
Your previous suggestion weren't tail recursion functions but this one is, so it looks not good to me.

> Regarding readability

I guess that although this is kind of subjective, recursion is usually conceived as a more elegant and readable solution than iteration, and even though I chose iteration for my first approach, I don't have anything against recursion (I actually enjoy designing recursive algorithms while keeping tail-recursion in mind)

> Regarding the usage of `i` in the iteration definition

True, I actually like the `for ; len(uuids) > 0; {` that you suggest, it is cleaner.

> In summary

Now that you took the time to do a suggestion for the tail-recursive function that keeps batch queries, tomorrow I'll extract this part of the code to a recursive function following your suggestion.
I may want to try to keep the happy just-one-query path for the cases where pods are given, which is also very go-stylish as it means an early return in the function. We can discuss it tomorrow.

Reply:
> True, I actually like the for ; len(uuids) > 0; { that you suggest, it is cleaner.

Actually, I just looked again and you can drop the semi-colons:

````
for len(uuids) > 0 {
        ...
}
```

Reply:
If you *really* want to optimize, you could consider using a stored procedure here instead. Although stored procedures and individual calls perform about the same, the risk of a network issue during the calls is significantly higher for individual calls rather than the one stored procedure call.   :-)

I am *not* proposing this change, just pointing out that it might actually be more performant.

Reply:
hmm, I am not against it if the performance isn't enough, but at first I believe that being able to receive intermediate results in the API may end up in delivering more information to the user of potential problems.

For now, I updated this logic from iteration to recursion as you suggested